[Merged by Bors] - Ultra Fast Super Slick CI

AgeManning opened the pull request:
Attempting to improve our CI speeds as its recently been a pain point.

Major changes:

 - Use a github action to pull stable/nightly rust rather than building it each run
 - Shift test suite to `nexttest` https://github.com/nextest-rs/nextest for CI
 
 UPDATE:

So I've iterated on some changes, and although I think its still not optimal I think this is a good base to start from. Some extra things in this PR:
- Shifted where we pull rust from. We're now using this thing: https://github.com/moonrepo/setup-rust . It's got some interesting cache's built in, but was not seeing the gains that Jimmy managed to get. In either case tho, it can pull rust, cargofmt, clippy, cargo nexttest all in < 5s. So I think it's worthwhile. 
- I've grouped a few of the check-like tests into a single test called `code-test`. Although we were using github runners in parallel which may be faster, it just seems wasteful. There were like 4-5 tests, where we would pull lighthouse, compile it, then run an action, like clippy, cargo-audit or fmt. I've grouped these into a single action, so we only compile lighthouse once, then in each step we run the checks. This avoids compiling lighthouse like 5 times.
- Ive made doppelganger tests run on our local machines to avoid pulling foundry, building and making lcli which are all now baked into the images. 
- We have sccache and do not incremental compile lighthouse

Misc bonus things:
- Cargo update
- Fix web3 signer openssl keys which is required after a cargo update
- Use mock_instant in an LRU cache test to avoid non-deterministic test
- Remove race condition in building web3signer tests

There's still some things we could improve on. Such as downloading the EF tests every run and the web3-signer binary, but I've left these to be out of scope of this PR. I think the above are meaningful improvements.

AgeManning committed:
First attempt will never work. Here's to hoping

AgeManning committed:
Correct simple mistakes

AgeManning committed:
Fix makefile

AgeManning committed:
More fixes

AgeManning committed:
Add configuration to nextest

antondlr committed:
detect self-hostedness through runner name

antondlr committed:
try different syntax for `if` statement

antondlr committed:
test latest windows runner image

paulhauner committed:
Don't run block processing tests in debug

AgeManning committed:
Looser name matching

AgeManning committed:
What is this env?

AgeManning committed:
Fix builder name matching

AgeManning committed:
Cleanup if statements

antondlr committed:
give us some `sccache` stats

AgeManning committed:
Trial a new rust builder

AgeManning committed:
Fix slasher tests

antondlr committed:
now actually show cache stats

jimmygchen commented on .github/workflows/test-suite.yml:
```diff
@@ -52,139 +54,191 @@ jobs:
+      uses: moonrepo/setup-rust@v1
+      with:
+          channel: stable
+          components: rustfmt
```
Do we need to install stable rust on hosted runners? they seem to update the image weekly, and is on latest Rust stable version as well

https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2204-Readme.md

I also noticed that `moonrepo/setup-rust@v1` caches the `debug` target by default, we could probably make use of it's caching capability (if it's comparable with `rust-cache` but may need `cache-target: release` for some jobs.

jimmygchen replied:
oh they've actually done a bit of comparison with `rust-cache` here: https://github.com/moonrepo/setup-rust#swatinemrust-cache

AgeManning replied:
Yeah I wasn't sure about updating the stable rust version. 

On actual runs, using this takes < 5s, and in some cases gives us cargo-nextest also. As you've pointed it out, it does misc caching. Figured I'd give it a try. I decided to use debug, as I assumed the debug would cache across runs, but perhaps I should cache release for release builds. 

The old way of getting stable rust is unmaintained, so this seems like the modern way to do it, and avoids misc errors. 

I'll update to cache some release builds and see how it goes.

antondlr committed:
proceed if folder exists

antondlr committed:
rename `mkdir` to avoid clash

AgeManning committed:
Update caching

AgeManning committed:
Improve workflow and correct web3 tests

AgeManning committed:
Use mock-instant for lru-time based tests

AgeManning committed:
Fix racey tests and doppleganger ci

antondlr committed:
show the stats forreal this time

antondlr committed:
show cache stats on windows

antondlr committed:
this should already be present in the runner
according to https://github.com/actions/runner-images/blob/ubuntu22/20230911.1/images/linux/Ubuntu2204-Readme.md
(closes version matches: go 1.20.8 .net 6.0.413)

AgeManning committed:
See how grouping runs goes

jimmygchen commented on Makefile:
```diff
@@ -108,11 +108,21 @@ build-release-tarballs:
+# Runs the full workspace tests in **release**, without downloading any additional
+# test vectors, using nextest.
+nextest-release:
+	cargo nextest run --workspace --release --exclude ef_tests --exclude beacon_chain --exclude slasher
```
Just reading through the docs in the default nextest config - for CI, it might be useful to run with `--no-fail-fast` and `--failure-output immediate-final`?

https://github.com/sigp/lighthouse/blob/ebd97bfa8390788dde4cd13382eca1fee4411c6a/.config/nextest.toml#L66-L68

https://github.com/sigp/lighthouse/blob/ebd97bfa8390788dde4cd13382eca1fee4411c6a/.config/nextest.toml#L49-L60

michaelsproul replied:
I think we might ideally want `fail-fast` for Bors and full testing for PRs. If we had to choose, I'd say fail-fast, because then we get feedback the quickest and free up the runner.

jimmygchen replied:
yeah good point  👍  given the amount of time our CI takes 🐢  (but soon 🐰!! )

AgeManning replied:
Yeah. I saw this too. I thought we wanted to end tests straight up and exit, for resource use. 

Often if there is a bug, usually i can find it without knowing all future test results

jimmygchen commented on .github/workflows/test-suite.yml:
```diff
@@ -46,145 +46,196 @@ jobs:
+      uses: moonrepo/setup-rust@v1
+      with:
+          channel: stable
+          components: rustfmt
```
It seems that we don't need `rustfmt` here, so could potentially get rid of this line, but it probably saves nothing (except a line of code) since this step completes in < 5s.

antondlr committed:
restrict `release-tests-windows` to CI box

antondlr committed:
additonal runner labels

antondlr committed:
disable cache for this task to preserve diskspace

jimmygchen commented on .github/workflows/test-suite.yml:
```diff
@@ -257,89 +324,65 @@ jobs:
-      run: cargo install cargo-udeps --locked --force
+    - name: Get latest version of nightly Rust
+      uses: moonrepo/setup-rust@v1
+      with:
+          channel: nightly
+          bins: cargo-udeps
```
Looking at the logs and looks like it's trying to locate `GITHUB_TOKEN` to use for downloading artifacts:

```
 WARN Failed to read git credential file /home/runner/.git-credentials err=Os { code: 2, kind: NotFound, message: "No such file or directory" }
 WARN Failed to read git credential file /home/runner/.config/git/credentials err=Os { code: 2, kind: NotFound, message: "No such file or directory" }
 WARN Failed to retrieve token from `gh auth token` err=Custom { kind: Other, error: "process exited with `exit status: 1`" }
```

I might be worth supplying the `GITHUB_TOKEN` to prevent getting rate limited by github:

jimmygchen replied:
Adding `GITHUB_TOKEN` and disable caching here.

Caching is probably not going to be useful for `cargo-udeps` and I think we can turn it off here:

```suggestion
      with:
          channel: nightly
          bins: cargo-udeps
          cache: false
      env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

jimmygchen commented on .github/workflows/test-suite.yml:
```diff
@@ -257,89 +324,65 @@ jobs:
     - name: Get latest version of stable Rust
-      run: rustup update stable
+      uses: moonrepo/setup-rust@v1
+      with:
+          channel: stable
+          cache-target: release
+          components: rustfmt
+          cache: false
```
Setting GITHUB_TOKEN to avoid rate limiting:

```suggestion
      with:
          channel: stable
          cache-target: release
          components: rustfmt
          cache: false
      env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

jimmygchen commented on .github/workflows/test-suite.yml:
```diff
@@ -46,145 +46,196 @@ jobs:
       if: env.SELF_HOSTED_RUNNERS == false
-      run: rustup update stable
+      uses: moonrepo/setup-rust@v1
+      with:
+          channel: stable
+          components: rustfmt
+          cache-target: release
+          bins: cargo-nextest
```
Setting GITHUB_TOKEN to avoid rate limiting and remove `rustfmt`:

```suggestion
      with:
          channel: stable
          cache-target: release
          bins: cargo-nextest
      env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

jimmygchen commented on .github/workflows/test-suite.yml:
```diff
@@ -46,145 +46,196 @@ jobs:
     steps:
     - uses: actions/checkout@v3
     - name: Get latest version of stable Rust
       if: env.SELF_HOSTED_RUNNERS == false
```
We should add single quotes to all of these comparisons, as `${{ }}` evaluates to string, which is why the comparison with `true` doesn't work 

```suggestion
      if: env.SELF_HOSTED_RUNNERS == 'false'
```

jimmygchen commented on .github/workflows/test-suite.yml:
```diff
@@ -46,145 +46,196 @@ jobs:
       if: env.SELF_HOSTED_RUNNERS == false
-      run: rustup update stable
+      uses: moonrepo/setup-rust@v1
+      with:
+          channel: stable
+          cache-target: release
+          components: rustfmt
+          bins: cargo-nextest
```
Setting GITHUB_TOKEN to avoid rate limiting and remove `rustfmt`:

```suggestion
      with:
          channel: stable
          cache-target: release
          bins: cargo-nextest
      env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

jimmygchen commented on .github/workflows/test-suite.yml:
```diff
@@ -46,145 +46,196 @@ jobs:
+          cache-target: release
+          bins: cargo-nextest
     - name: Install Foundry (anvil)
+      if: env.SELF_HOSTED_RUNNERS == false
```
wrapping in single quotes for string comparison:

```suggestion
      if: env.SELF_HOSTED_RUNNERS == 'false'
```

jimmygchen commented on .github/workflows/test-suite.yml:
```diff
@@ -46,145 +46,196 @@ jobs:
+          components: rustfmt
+          bins: cargo-nextest
     - name: Install Foundry (anvil)
+      if: env.SELF_HOSTED_RUNNERS == false
```
wrapping in single quotes for string comparison:

```suggestion
      if: env.SELF_HOSTED_RUNNERS == 'false'
```

jimmygchen commented on .github/workflows/test-suite.yml:
```diff
@@ -231,21 +291,28 @@ jobs:
+          cache-target: release
+          components: rustfmt
     - name: Install geth
+      if: env.SELF_HOSTED_RUNNERS == false
```
wrapping in single quotes for string comparison:

```suggestion
      if: env.SELF_HOSTED_RUNNERS == 'false'
```

jimmygchen commented on .github/workflows/test-suite.yml:
```diff
@@ -257,89 +324,65 @@ jobs:
     name: check-msrv
     runs-on: ubuntu-latest
-    needs: [cargo-fmt, extract-msrv]
+    needs: [extract-msrv]
```
nitpick: we can probably move the steps from `extract-msrv` to this job as the output isn't used anywhere else. This saves us from one extra git checkout, so probably not super impactful, happy to do this in a separate PR if you prefer.

jimmygchen commented on .github/workflows/test-suite.yml:
```diff
@@ -257,89 +324,65 @@ jobs:
-      run: rustup update stable
+      uses: moonrepo/setup-rust@v1
+      env:
+          CARGO_INCREMENTAL: 1
```
Is this intended for this step? Not sure if it compiles anything here

AgeManning replied:
My thinking was that at each step we compile lighthouse and its deps to do the all the various checks. I thought incremental compilation would be beneficial here. This isn't the case? @jimmygchen

jimmygchen replied:
Oh it makes sense - in that case i think we need to shift this `env` one level up to the _job_ level, e.g. 

```
  check-code:
    name: check-code
    runs-on: ubuntu-latest
    env:
        CARGO_INCREMENTAL: 1
```

At the moment it's scoped at the _step_ level, which i don't think is available to other steps under this job.

jimmygchen commented on .github/workflows/test-suite.yml:
```diff
@@ -46,145 +46,196 @@ jobs:
     - name: Install make
-      run: choco install -y make
-    - uses: KyleMayes/install-llvm-action@v1
       if: env.SELF_HOSTED_RUNNERS == false
```
wrapping in single quotes for string comparison:

```suggestion
      if: env.SELF_HOSTED_RUNNERS == 'false'
```

jimmygchen commented on .github/workflows/test-suite.yml:
```diff
@@ -46,145 +46,196 @@ jobs:
     - name: Set LIBCLANG_PATH
       run: echo "LIBCLANG_PATH=$((gcm clang).source -replace "clang.exe")" >> $env:GITHUB_ENV
+    - name: Install nextest for windows
+      if: env.SELF_HOSTED_RUNNERS == false
```
wrapping in single quotes for string comparison:

```suggestion
      if: env.SELF_HOSTED_RUNNERS == 'false'
```

jimmygchen commented on .github/workflows/test-suite.yml:
```diff
@@ -46,145 +46,196 @@ jobs:
     steps:
     - uses: actions/checkout@v3
     - name: Get latest version of stable Rust
       if: env.SELF_HOSTED_RUNNERS == false
```
wrapping in single quotes for string comparison:

```suggestion
      if: env.SELF_HOSTED_RUNNERS == 'false'
```

jimmygchen commented on .github/workflows/test-suite.yml:
```diff
@@ -46,145 +46,196 @@ jobs:
+          components: rustfmt
+          bins: cargo-nextest
     - name: Install Foundry (anvil)
+      if: env.SELF_HOSTED_RUNNERS == false
```
wrapping in single quotes for string comparison:

```suggestion
      if: env.SELF_HOSTED_RUNNERS == 'false'
```

jimmygchen commented on .github/workflows/test-suite.yml:
```diff
@@ -46,145 +46,196 @@ jobs:
+      with:
+          channel: stable
+          cache-target: release
+          components: rustfmt
```
```suggestion
```

jimmygchen commented on .github/workflows/test-suite.yml:
```diff
@@ -46,145 +46,196 @@ jobs:
+      with:
+          channel: stable
+          cache-target: release
+          components: rustfmt
```
Looks like we don't need Rust for the docker build.

jimmygchen commented on .github/workflows/test-suite.yml:
```diff
@@ -46,145 +46,196 @@ jobs:
+      with:
+          channel: stable
+          cache-target: release
+          components: rustfmt
```
```suggestion
```

jimmygchen commented on .github/workflows/test-suite.yml:
```diff
@@ -194,11 +245,14 @@ jobs:
+      with:
+          channel: stable
+          cache-target: release
+          components: rustfmt
```
```suggestion
```

jimmygchen commented on .github/workflows/test-suite.yml:
```diff
@@ -208,21 +262,27 @@ jobs:
+      with:
+          channel: stable
+          cache-target: release
+          components: rustfmt
```
```suggestion
```

jimmygchen commented on .github/workflows/test-suite.yml:
```diff
@@ -208,21 +262,27 @@ jobs:
+      with:
+          channel: stable
+          cache-target: release
+          components: rustfmt
```
```suggestion
```

jimmygchen commented on .github/workflows/test-suite.yml:
```diff
@@ -231,21 +291,28 @@ jobs:
     - uses: actions/checkout@v3
     - name: Get latest version of stable Rust
-      run: rustup update stable
+      if: env.SELF_HOSTED_RUNNERS == false
```
wrapping in single quotes for string comparison:

```suggestion
      if: env.SELF_HOSTED_RUNNERS == 'false'
```

jimmygchen commented on .github/workflows/test-suite.yml:
```diff
@@ -231,21 +291,28 @@ jobs:
+      with:
+          channel: stable
+          cache-target: release
+          components: rustfmt
```
```suggestion
```

jimmygchen commented on .github/workflows/test-suite.yml:
```diff
@@ -231,21 +291,28 @@ jobs:
-          make install-lcli
+    - name: Install lcli
+      if: env.SELF_HOSTED_RUNNERS == false
+      run: make install-lcli
```
We lose the check for compatibility between testnet & `lcli` here (that could potentially catch `lcli` breakage), so we might get feedback for this a week late - but I'm sure this has been considered, and I don't mind it if we can get good gains from this.

jimmygchen replied:
There's also a chance that this test would keep failing  if there's changes that make `lcli` incompatible with the latest BN, and we'd have to update the runner image to the get test to pass.

AgeManning replied:
Yep, we have considered this. We can add it back, but we figured we'd leave it out until it causes pain. Its not often we have these incompatibilities, i think (famous last words)

jimmygchen replied:
Sounds good leaving it until it causes pain 👍

jimmygchen commented on .github/workflows/test-suite.yml:
```diff
@@ -257,89 +324,65 @@ jobs:
+      with:
+          channel: stable
+          cache-target: release
+          components: rustfmt
```
Sorry to be annoying here, thought I might as well keep doing this since I've  already started 😅 

```suggestion
```

jimmygchen commented on .github/workflows/test-suite.yml:
```diff
@@ -46,145 +46,196 @@ jobs:
       if: env.SELF_HOSTED_RUNNERS == false
-      run: rustup update stable
+      uses: moonrepo/setup-rust@v1
+      with:
+          channel: stable
+          cache-target: release
+          components: rustfmt
+          bins: cargo-nextest
```
Setting GITHUB_TOKEN to avoid rate limiting and remove `rustfmt`:

```suggestion
      with:
          channel: stable
          cache-target: release
          bins: cargo-nextest
      env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

jimmygchen commented on .github/workflows/test-suite.yml:
```diff
@@ -46,145 +46,196 @@ jobs:
     - name: Get latest version of stable Rust
-      run: rustup update stable
+      uses: moonrepo/setup-rust@v1
+      with:
+          channel: stable
+          cache-target: release
+          components: rustfmt
+          bins: cargo-nextest
```
Setting GITHUB_TOKEN to avoid rate limiting and remove `rustfmt`:

```suggestion
      with:
          channel: stable
          cache-target: release
          bins: cargo-nextest
      env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

jimmygchen commented on .github/workflows/test-suite.yml:
```diff
@@ -46,145 +46,196 @@ jobs:
     - name: Get latest version of stable Rust
-      run: rustup update stable
+      uses: moonrepo/setup-rust@v1
+      with:
+          channel: stable
+          cache-target: release
+          components: rustfmt
+          bins: cargo-nextest
```
Setting GITHUB_TOKEN to avoid rate limiting and remove `rustfmt`:

```suggestion
      with:
          channel: stable
          cache-target: release
          bins: cargo-nextest
      env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

jimmygchen commented on .github/workflows/test-suite.yml:
```diff
@@ -46,145 +46,196 @@ jobs:
       if: env.SELF_HOSTED_RUNNERS == false
-      run: rustup update stable
+      uses: moonrepo/setup-rust@v1
+      with:
+          channel: stable
+          components: rustfmt
+          bins: cargo-nextest
```
Setting GITHUB_TOKEN to avoid rate limiting and remove `rustfmt`:

```suggestion
      with:
          channel: stable
          bins: cargo-nextest
      env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

jimmygchen reviewed (changes requested):
Niiice, we're close! our CI is looking great 🔥 

I've added lots of comments but mostly the same thing (Sorry I thought it would make it easier to batch commit the suggestions, but there are too many occurences!)

So the main things are
- use `GITHUB_TOKEN` for `setup-rust` to avoid getting rate limited (we were getting this quite often before)
- wrap the `env.SELF_HOSTED_RUNNERS` booleans with single quotes as `${{ }}` evaluates to string

Others are just nitpicks, feel free to leave them out ☺️

michaelsproul commented on testing/web3signer_tests/tls/lighthouse/cert.pem:
```diff
@@ -1,32 +1,33 @@
 -----BEGIN CERTIFICATE-----
```
Ah, I also just re-generated these in #4650. You'll get a merge conflict but I don't mind if you just clobber the keys I generated. Explicitly specifying aes256 is probably better than what I did (relying on the default).

AgeManning committed:
Update .github/workflows/test-suite.yml

Co-authored-by: Jimmy Chen <jchen.tc@gmail.com>

AgeManning committed:
Update .github/workflows/test-suite.yml

Co-authored-by: Jimmy Chen <jchen.tc@gmail.com>

AgeManning committed:
Update .github/workflows/test-suite.yml

Co-authored-by: Jimmy Chen <jchen.tc@gmail.com>

AgeManning committed:
Update .github/workflows/test-suite.yml

Co-authored-by: Jimmy Chen <jchen.tc@gmail.com>

AgeManning committed:
Update .github/workflows/test-suite.yml

Co-authored-by: Jimmy Chen <jchen.tc@gmail.com>

AgeManning committed:
Update .github/workflows/test-suite.yml

Co-authored-by: Jimmy Chen <jchen.tc@gmail.com>

AgeManning committed:
Apply reviewers comments

jimmygchen commented on .github/workflows/test-suite.yml:
```diff
@@ -34,157 +34,195 @@ jobs:
+#        directory: ${{ runner.temp }}/llvm
     - name: Set LIBCLANG_PATH
       run: echo "LIBCLANG_PATH=$((gcm clang).source -replace "clang.exe")" >> $env:GITHUB_ENV
+    - name: Install nextest for windows
+      if: env.SELF_HOSTED_RUNNERS == 'false'
+      run: |
+        powershell Invoke-WebRequest -Uri https://get.nexte.st/latest/windows -OutFile nextest-windows.zip &&
+        powershell Expand-Archive -Path 'nextest-windows.zip' -DestinationPath 'C:\Users\ContainerAdministrator\.cargo\bin\'
```
Do we still need this now the `moonrepo/setup-rust` step installs the `cargo-nextest` binary?

AgeManning replied:
Removed this

AgeManning committed:
Move env variable to job level and remove windows manual nextest install

AgeManning committed:
Merge latest unstable

AgeManning committed:
Update workflow

jimmygchen commented on Cargo.lock:
Is deleting this file intentional?

AgeManning committed:
Update lockfile and msrv CI

AgeManning committed:
Re-committing lockfile which didn't get included last run

jimmygchen reviewed (approved):
Looks good to me! Can't wait to get this in 🚀

paulhauner commented:
bors r+

paulhauner commented:
bors r-

paulhauner commented:
bors r+

paulhauner committed:
Clean `bors.toml`

paulhauner committed:
Clean `bors.toml`

paulhauner commented:
bors r-

paulhauner commented:
bors r+

realbigsean committed:
remove yet-to-be-added job

AgeManning committed:
Merge #4795

paulhauner commented:
bors r+

AgeManning committed:
Merge unstable

AgeManning commented:
bors retry

paulhauner commented:
bors retry

paulhauner commented:
bors retry
